arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedColumnIterator.java

```diff
@@ -69,12 +69,14 @@ public boolean producesDictionaryEncodedVector() {
   }
 
   public abstract class BatchReader {
-    public void nextBatch(FieldVector fieldVector, int typeWidth, NullabilityHolder holder) {
+    public void nextBatch(
+        int numValsToRead, FieldVector fieldVector, int typeWidth, NullabilityHolder holder) {
       int rowsReadSoFar = 0;
-      while (rowsReadSoFar < batchSize && hasNext()) {
+      while (rowsReadSoFar < batchSize && hasNext() && rowsReadSoFar < numValsToRead) {
         advance();
+        int expectedBatchSize = Math.min(batchSize - rowsReadSoFar, numValsToRead - rowsReadSoFar);
```

Comment:
We definitely need to guard num_vals to read here to be greater than -1, we still have the potential for overflow but probably not huge if we require -1 as the only negative value allowed

Reply:
I added a check to use the original logic `batchSize - rowsReadSoFar` if `numValsToRead` is less than zero.
```
        if (numValsToRead < 0) {
          expectedBatchSize = batchSize - rowsReadSoFar;
        }
```